direct_consumer: locking and no leader fixes

joe-redpanda opened the pull request:
Direct consumer previously locked update_fetchers, but this lock was not guarding against concurrent modification of its _subscriptions map.

This commit changes update_fetchers to demand lock units from all callers, and return those lock units on completion. All updating apis now aquire a lock before modifying the underlying _subscriptions map.

Additionally adds proper leader assignment for when no leader is found. Previously the leader would not get reset to 'nullopt', which lead to the assigned fetch offset being overwritten with nullopt which inappropriately reset the fetch offset to the result of list_offsets

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [x] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
### Bug Fixes

* stronger locking guarantees on direct_consumer assignments, prevent iterator invalidation

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

joe-redpanda committed:
direct_consumer: stronger locking

Direct consumer previously locked update_fetchers, but this lock was not
guarding against concurrent modification of its _subscriptions map.

This commit changes update_fetchers to demand lock units from all
callers, and return those lock units on completion. All updating apis
now aquire a lock before modifying the underlying _subscriptions map.

dotnwat commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -223,11 +235,13 @@ ss::future<> direct_consumer::unassign_partitions(
 ss::future<> direct_consumer::handle_metadata_update() {
-    co_await update_fetchers();
+    auto lock_holder = co_await _subscriptions_lock.get_units();
+    std::ignore = co_await update_fetchers(std::move(lock_holder));
```
AFAICT all callers to `update_fetchers` ignore their return value, so why return the units at all?

joe-redpanda replied:
At the time of writing I was considering having an api like

foo() {
holder = get_units()

// do something

holder = co_await update_fetcher(holder)

// do some additional things whilst still locked

}

If I havent found a use for the retuned units by the time I've shaken some current debugging out I'll drop the return units

joe-redpanda replied:
Cleaned up to unblock bharath

joe-redpanda committed:
direct_consumer: no leader fix

The fetcher assignment's leader needs to be updated to nullopt if there
is no current leader. This prevents the next update loop from
overwriting the fetch offset with nullopt.

joe-redpanda commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -15,6 +15,8 @@
 #include "model/validation.h"
 #include "ssx/future-util.h"
 
+#include <tuple>
```
not sure how this slipped in

bharathv commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -50,14 +50,14 @@ void direct_consumer::update_configuration(configuration cfg) {
-ss::future<>
-direct_consumer::update_fetchers(topic_partition_map<subscription> removals) {
+ss::future<> direct_consumer::update_fetchers(
+  [[maybe_unused]] mutex::units lock_holder,
```
hmm.. I'd like to see @mmaslankaprv take on this commit..

> guarding against concurrent modification of its _subscriptions map.

The fix makes sense to me, I think the issue is update_fetchers() loops through _subscriptions with scheduling points and a concurrent unassign may just invalidate the iterators?

If that is the case I think that should have more unit tests that can catch this invalidation?

The issue is this serializes all assign/unassign requests but maybe thats ok.

bharathv replied:
ok to add tests in future patches.

joe-redpanda replied:
awesome, that was going to be my ask

bharathv commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -85,6 +85,7 @@ ss::future<> direct_consumer::update_fetchers(
                       model::topic_partition_view(topic, p_id));
                     // preserve the fetch offset for the next fetcher to use
                     sub.fetch_offset = offset;
+                    sub.current_fetcher = std::nullopt;
```
nice find 👍

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--012c0778-faaf-4258-b5ea-47854d0a7443-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72208">build#72208</a></summary>


| test_class         | test_method                | test_arguments         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-------------------|:---------------------------|:-----------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogTestOauth  | test_kafka_oauth           | {"authz_match": "acl"} | integration | https://buildkite.com/redpanda/redpanda/builds/72208#0199403b-59d9-402f-ac22-108e629949a0 | FLAKY         | 20/21    | upstream reliability is '98.7878787878788'. current run reliability is '95.23809523809523'. drift is 3.54978 and the allowed drift is set to 50. The test should PASS   |
| SimpleEndToEndTest | test_consumer_interruption | null                   | integration | https://buildkite.com/redpanda/redpanda/builds/72208#0199403e-28b1-4bb8-a0ec-92e6e5915964 | FLAKY         | 17/21    | upstream reliability is '99.28571428571429'. current run reliability is '80.95238095238095'. drift is 18.33333 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

mmaslankaprv commented:
Good find
